### PR TITLE
pkg/reloader: improve detection of directory changes

### PR DESCRIPTION
**Pull request by simonpasquier**
* [X] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

When watching for changes in directories, the reloader used to rely only
on the watch interval and not on the inotify events. This commit implements
a more efficient detection of changes for watched directories.

The change also adds a new `DelayInterval` option that allows to delay
the config reload after no additional event are received.

Finally a new metric,
`thanos_sidecar_reloader_config_apply_operations_total`, is added and
`thanos_sidecar_reloader_config_apply_errors_total` has been renamed
to `thanos_sidecar_reloader_config_apply_operations_failed_total` for
consistency.

## Verification

I've modified the unit test to validate that the reloader reacts on file changes (without the periodic watch). I've added a unit test to validate that the reloader's periodic watch detects a change in case it hasn't been caught by fsnotify.

I've also included this PR in https://github.com/prometheus-operator/prometheus-operator/pull/3457 which is the change that triggered this PR.

### CODE REVIEW

**Comment by kakkoyun** [comment]
@simonpasquier will this have any effect on falky reloader tests https://github.com/thanos-io/thanos/issues/2622? 

Also e2e tests are failing.

**Comment by simonpasquier** [comment]
Kemal, thanks for pointing the issue, I'll look into it.

File: pkg/reloader/reloader.go
Comment (kakkoyun) [inline pkg/reloader/reloader.go:224]: nit: Renaming the `cancel` to `applyCancel` could make it easier to track.
```suggestion
	applyCtx, applyCancel := context.WithTimeout(ctx, r.watchInterval)
```

File: pkg/reloader/reloader.go
Comment (kakkoyun) [inline pkg/reloader/reloader.go:474]: nit: Does it ever get called concurrently? If it's the case we could use `sync.Once` for initialization.

File: pkg/reloader/reloader.go
Comment (kakkoyun) [inline pkg/reloader/reloader.go:506]: nit
```suggestion
	var (
		wg sync.WaitGroup
		notify = make(chan struct{})
	)
```

**Review by kakkoyun — APPROVED** [review]
Amazing work. LGTM AFAICT.

**Comment by kakkoyun** [comment]
Let's have another pair of eyes before we merge this.  cc @thanos-io/thanos-maintainers

File: pkg/reloader/reloader.go
Comment (simonpasquier) [reply pkg/reloader/reloader.go:474]: No, `addPath` and `addDir` are called sequentially in `*Reloader.Watch()`.

File: pkg/reloader/reloader.go
Comment (simonpasquier) [inline pkg/reloader/reloader.go:188]: oops this needs to be removed.

**Comment by simonpasquier** [comment]
prometheus-operator/prometheus-operator#3457 is the prometheus-operator PR including this change.

I'm not sure that this PR will help to solve https://github.com/thanos-io/thanos/issues/2622. @kakkoyun do you have a recent CI failure for this? I've run the unit tests several times (>100) on my machine with `-race -cpu 2` and all passed (both on `master` and with the PR branch).

**Comment by kakkoyun** [comment]
Thanks a lot for addressing issues @simonpasquier 💜 The flakiness happens on a single core and relatively slow CPU setups (such as CI). It's ok, it's not our priority for this PR. Let's see if it helps after merging this.

**Review by brancz — APPROVED** [review]
Nice!

**Pull request merged**
